pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,459 @@
+)
+
+// traceValue represents the trace information for a single registration.
+type traceValue struct {
+	traced kvserverpb.TracedEntry
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+
+	mu struct {
+		syncutil.Mutex
+
+		// seenMsgAppResp tracks whether a MsgAppResp message has already been
+		// logged by each replica peer. This limits the size of the log at a
+		// small risk of missing some important messages in the case of dropped
+		// messages or reproposals.
+		seenMsgAppResp map[raftpb.PeerID]bool
+
+		// seenMsgStorageAppendResp tracks whether a MsgStorageAppendResp
+		// message has already been logged.
+		seenMsgStorageAppendResp bool
+
+		// propCtx is the underlying proposal context used for tracing to the
+		// SQL trace.
+		propCtx context.Context
+
+		// propSpan is the span connected to the propCtx. It must be finished
+		// when the trace is removed.
+		propSpan *tracing.Span
+	}
+}
+
+// logf logs the message to the trace context and the proposal context. The
+// proposal context is populated on the leaseholder and is attached to the SQL
+// trace.
+func (t *traceValue) logf(depth int, format string, args ...interface{}) {
+	log.InfofDepth(t.ctx, depth+1, format, args...)
+
+	t.mu.Lock()
+	propCtx := t.mu.propCtx
+	t.mu.Unlock()
+	if propCtx != nil {
+		log.VEventfDepth(propCtx, depth+1, 3, format, args...)
+	}
+}
+
+// sawMsgAppResp returns true on the first MsgAppResp for this peer.
+func (t *traceValue) sawMsgAppResp(p raftpb.PeerID) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.mu.seenMsgAppResp[p] {
+		return false
+	}
+	t.mu.seenMsgAppResp[p] = true
+	return true
+}
+
+// sawMsgStorageAppendResp returns true on the first MsgStorageAppendResp.
+func (t *traceValue) sawMsgStorageAppendResp() bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.mu.seenMsgStorageAppendResp {
+		return false
+	}
+	t.mu.seenMsgStorageAppendResp = true
+	return true
+}
+
+// String attempts to balance uniqueness with readability by only keeping the
+// lower 16 bits of the trace and span.
+func (tv *traceValue) String() string {
+	return redact.StringWithoutMarkers(tv)
+}
+
+func (tv *traceValue) SafeFormat(w redact.SafePrinter, _ rune) {
+	w.Printf("i%d/%x.%x", tv.traced.Index, uint16(tv.traced.TraceID), uint16(tv.traced.SpanID))
+}
+
+// RaftTracer is a utility to trace the lifetime of raft log entries. It may
+// include some unrelated entries, since it does not consider the term. It
+// traces at most one MsgAppResp and MsgStorageAppendResp per index which is the
+// first one that is past our index entry. This limitation means it may not
+// capture all the relevant messages particularly if the term changes.
+//
+// The library will log in two different ways once to the standard cockroach log
+// and once to the SQL trace on the leaseholder.
+// TODO(baptist): Look at logging traces on followers and sending back to the
+// leader. It would need to be best effort, but might still be useful.
```

Comment:
Alternatively, double-down on distributed trace collection if/when it's supported. So that the trace does not need to be plumbed back to the leaseholder / txn coordinator.

Reply:
Added to TODO.